### PR TITLE
Ports - Ngoc

### PR DESCRIPTION
**Pull request by lebaongoc**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Model contains data for the application and the state of the application
Describe in your own words what the Controller is doing in Rails | The Controller receive HTTP requests from the outside world, gather data from the model and pass dada to the view to be rendered as HTML
Describe in your own words what the View is doing in Rails | Display data in HTML
Describe an edge-case controller test you wrote | For the update function, I wrote a test to check if the Controller redirect user back to the root page if an invalid task id was provided from the user
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params is used to provide a safe way to require and permit data that comes from our forms.
How are Rails migrations related to Rails models? | Rails migrations provide the documentation and instructions to the database on how to store the data the model is describing
Describe one area of Rails that are still unclear on | When I use resources to generates the default RESTful routes. I ran into error with the link to create new task on the main page (index.html.erb-last line of code)- the error was that the route was not found. I fixed this by adding " method: :get" to specify the type of request with this link. However, I did not need to do this before implementing 'resources'. Was this caused by the order of the RESTful routes that resources provided?

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  </head>
+
+  <body>
+    
+    <main>
+      <%= yield %>
+    </main>
+
+    <footer>
+      <p>© 2019 Ngoc Le </p>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:19]: 🤣

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -79,50 +76,104 @@
...
+      #task will get redirect and redirect to tasks_path
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
   end
 
-  # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "changes task's completion status from true to false" do
```
Comment (CheezItMan) [inline test/controllers/tasks_controller_test.rb:169]: You should also test that you can toggle a task back, after marking it complete.

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    else
+      task.destroy
+      redirect_to tasks_path
+    end
+  end
+
+  def mark_complete
+    task = Task.find_by(id: params[:id])
+
+    task.completed?
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:63]: What is this doing here?

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    if task.nil?
+      head :not_found
+    else
+      task.destroy
+      redirect_to tasks_path
+    end
+  end
+
+  def mark_complete
+    task = Task.find_by(id: params[:id])
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:61]: You also should have an if statement to ensure that you find the task before trying to toggle it.

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  def edit
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to tasks_path
+    end
+  end
+
+  def update
+    task = Task.find_by(id: params[:id])
+    task.update(task_params)
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:40]: You should have some check to ensure that you did find the task before doing update.

File: app/views/tasks/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+                <%= link_to "Delete Task", task_path(task.id), method: :delete, data: {confirm: "Are you sure?"} %>
+                
+            </section>
+
+        </li>
+    <% end %>
+</ul>
+
+<%= button_to "Create New Task ", new_task_path, method: :get, :class => "create-button" %>
```
Comment (CheezItMan) [inline app/views/tasks/index.html.erb:25]: The reason you had to specify method: :get in this line is because `button_to` creates a mini-form and by default all forms submit as a POST request.

File: app/views/tasks/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+<ul class="task-list">
+    <% @tasks.each do|task|%>
+        <li>
+        <%= link_to task.name, task_path(task.id), :class => "task-link"%>
+
+            <section class="index-button">
+
+                <% if task.completed == false %>
+                    <%= link_to "Mark Complete", {action: "mark_complete", id: task.id}, method: :patch %> |
```
Comment (CheezItMan) [inline app/views/tasks/index.html.erb:12]: You should use paths instead of specifying the controller actions like this.

```erb
<%= link_to "Mark Complete", mark_complete_path(task.id) %>
```

**Comment by CheezItMan** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | Good use of git, more commits would be better.
Answered comprehension questions | Check, see my inline comment on your question.
Successfully handles: Index, Show |  Check
Index & Show tests pass |Check
Successfully handles: New, Create | Check
New & Create tests pass |Check
Successfully handles: Edit, Update |Check
Tests for Edit & Update test valid & invalid task ids | Check
Successfully handles: Destroy, Task Complete | Check
Tests for Destroy & Task Complete include tests for valid and invalid task ids | Check
Routes follow RESTful conventions | Check
Uses named routes (like `_path`) | Mostly, but in a few areas you used another syntax that's frowned upon. 
**Overall** | Nice work, you hit all the learning goals. I like the tests you added, well done.  See my inline comments and let me know any questions you have.
